libs/wingsdk/src/target-tf-gcp/function.ts

```diff
@@ -0,0 +1,127 @@
+  case: CaseConventions.LOWERCASE,
+};
+
+const BUCKET_NAME_OPTS: NameOptions = {
+  maxLen: 54,
+  case: CaseConventions.LOWERCASE,
+  disallowedRegex: /([^a-z0-9_\-]+)/g,
+  includeHash: false,
+};
+/**
+ * GCP implementation of `cloud.Function`.
+ *
+ * @inflight `@winglang/wingsdk.cloud.IFunctionClient`
+ */
+
+export class Function extends cloud.Function {
+  private readonly function: CloudfunctionsFunction;
+  private readonly bucket: StorageBucket;
+  private readonly bucketObject: StorageBucketObject;
+
+  constructor(
+    scope: Construct,
+    id: string,
+    inflight: cloud.IFunctionHandler,
+    props: cloud.FunctionProps = {}
+  ) {
+    super(scope, id, inflight, props);
+
+    // app is a property of the `cloud.Function` class
+    const app = App.of(this) as App;
+
+    // bundled code is guaranteed to be in a fresh directory
+    const bundle = createBundle(this.entrypoint);
+
+    // Create Cloud Function executable
+    const asset = new TerraformAsset(this, "Asset", {
+      path: resolve(bundle.directory),
+      type: AssetType.ARCHIVE,
+    });
+
+    // create a bucket with unique name
+    const bucketName = ResourceNames.generateName(this, BUCKET_NAME_OPTS);
+    // GCP bucket names must be globally unique, but the Terraform resource
+    // provider doesn't provide a mechanism like `bucketPrefix` as AWS does,
+    // so we must generate a random string to append to the bucket name.
+
+    // The random string must be managed in Terraform state so that it doesn't
+    // change on every subsequent compile or deployment.
+    const randomId = new Id(this, "Id", {
+      byteLength: 4, // 4 bytes = 8 hex characters
+    });
+
+    // create the bucket
+    this.bucket = new StorageBucket(this, "Function_Bucket", {
+      name: bucketName + "-" + randomId.hex,
+      location: app.storageLocation,
+    });
+
+    // put the executable in the bucket as an object
+    this.bucketObject = new StorageBucketObject(
+      this,
+      "Function_Object_Bucket",
+      {
+        name: "objects",
+        bucket: this.bucket.name,
+        source: asset.path,
+      }
+    );
+
+    // create the cloud function
+    this.function = new CloudfunctionsFunction(this, "Default_Function", {
+      name: ResourceNames.generateName(this, FUNCTION_NAME_OPTS),
+      description: "This function was created by Wing",
+      runtime: "nodejs16",
+      availableMemoryMb: props.memory ?? 128,
+      sourceArchiveBucket: this.bucket.name,
+      sourceArchiveObject: this.bucketObject.name,
+      entryPoint: "handler",
+      triggerHttp: true,
+      timeout: props.timeout?.seconds ?? 60,
+      environmentVariables: props.env ?? {},
+    });
+
+    // create the permissions
+    new CloudfunctionsFunctionIamMember(this, "Function_IAM_Member", {
+      project: app.projectId,
+      region: app.storageLocation,
+      cloudFunction: this.function.name,
+      role: "roles/cloudfunctions.invoker",
+      member: "allUsers",
```

Comment:
I'm not really familiar with GCP, what does `allUsers` in this context mean?

Reply:
> I'm not really familiar with GCP, what does `allUsers` in this context mean?

This is for public function and IAM entry for all users to invoke the function

For example if we do this
`member = "user:myFunctionInvoker@example.com"`
this will only allow a specific user to have the permission to invoke

You can take a look at this tf doc
[Public Access](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/cloudfunctions_function.html#example-usage---public-function)
[Single User Access](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/cloudfunctions_function.html#example-usage---single-user)

Reply:
We do not want to allow all users to invoke the function. Permissions to the function should be granted during the bind method.